src/communication/v2/listener.hpp

```diff
@@ -0,0 +1,141 @@
+  ~Listener() {}
+
+  template <typename... Args>
+  static std::shared_ptr<Listener> Create(Args &&...args) {
+    return std::shared_ptr<Listener>{new Listener(std::forward<Args>(args)...)};
+  }
+
+  void Start() { DoAccept(); }
+
+  bool IsRunning() const noexcept { return alive_.load(std::memory_order_relaxed); }
+
+ private:
+  Listener(IOContextThreadPool &io_thread_context_pool, TSessionData *data, ServerContext *server_context,
+           tcp::endpoint &endpoint, const std::string_view service_name, const int inactivity_timeout_sec)
+      : io_thread_context_pool_(io_thread_context_pool),
+        data_(data),
+        server_context_(server_context),
+        acceptor_(io_thread_context_pool.GetIOContext()),
+        endpoint_{endpoint},
+        service_name_{service_name},
+        inactivity_timeout_sec_{inactivity_timeout_sec} {
+    boost::system::error_code ec;
+    // Open the acceptor
+    acceptor_.open(endpoint.protocol(), ec);
+    if (ec) {
+      OnError(ec, "open");
+      return;
+    }
+
+    // Allow address reuse
+    acceptor_.set_option(boost::asio::socket_base::reuse_address(true), ec);
+    if (ec) {
+      OnError(ec, "set_option");
+      return;
+    }
+
+    // Bind to the server address
+    acceptor_.bind(endpoint, ec);
+    if (ec) {
+      spdlog::error(
+          utils::MessageWithLink("Cannot bind to socket on endpoint {}.", endpoint, "https://memgr.ph/socket"));
+      OnError(ec, "bind");
+      return;
+    }
+
+    acceptor_.listen(boost::asio::socket_base::max_listen_connections, ec);
+    if (ec) {
+      OnError(ec, "listen");
+      return;
+    }
+  }
+
+  void DoAccept() {
+    acceptor_.async_accept(io_thread_context_pool_.GetIOContext(),
+                           [shared_this = this->shared_from_this()](auto ec, auto &&socket) {
+                             shared_this->OnAccept(ec, std::forward<decltype(socket)>(socket));
+                           });
+  }
+
+  void OnAccept(boost::system::error_code ec, tcp::socket socket) {
+    if (ec) {
+      return OnError(ec, "accept");
+    }
+
+    auto session = SessionHandler::Create(std::move(socket), this->data_, *this->server_context_, this->endpoint_,
+                                          this->inactivity_timeout_sec_, service_name_);
+    sessions_.WithLock([session = session](auto &sessions) {
+      // Clean disconnected clients
+      std::erase_if(sessions, [](const auto &elem) { return !elem->IsConnected(); });
+      sessions.insert(std::move(session));
+    });
+
+    session->Start();
+    DoAccept();
+  }
+
+  void OnError(const boost::system::error_code &ec, const std::string_view what) {
+    spdlog::error("Listener failed on {}: {}", what, ec.message());
+    alive_.store(false, std::memory_order_relaxed);
+  }
+
+  IOContextThreadPool &io_thread_context_pool_;
+  TSessionData *data_;
+  ServerContext *server_context_;
+  tcp::acceptor acceptor_;
+
+  tcp::endpoint endpoint_;
+  std::string_view service_name_;
```

Comment:
```suggestion
  std::string service_name_;
```

Reply:
Since service name is owned by the listener, I think it should also own the `service_name` instead of a view.

Reply:
Aha you are right nobody owns it, I will move it to the server since it makes more sense that the server owns it than the listener, since they both need it.